### ISSUE TITLE
Version 8 packages are not declared as es module

### ISSUE THREAD

**Issue by LennartH** [nartc/mapper#446]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the issue

I'm using `core` and `classes` in an es module project. When running a script I got the following error:

```
import { AutoMap } from '@automapper/classes';
         ^^^^^^^
SyntaxError: Named export 'AutoMap' not found. The requested module '@automapper/classes' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from '@automapper/classes';
const { AutoMap } = pkg;
```

After trying to reexport the necessary parts using the described default export I got a more descriptive error:

```
(node:28077) Warning: To load an ES module, set "type": "module" in the package.json or use the .mjs extension.
(Use `node --trace-warnings ...` to show where the warning was created)
.../node_modules/@automapper/core/index.esm.js:1570
export { CamelCaseNamingConvention, PascalCaseNamingConvention, SnakeCaseNamingConvention, addProfile, afterMap, beforeMap, condition, constructUsing, convertUsing, createMap, createMapper, defaultSerializerOptions, extend, forMember, forSelf, fromValue, getRecursiveValue, ignore, isDateConstructor, isEmpty, isPrimitiveConstructor, mapDefer, mapFrom, mapWith, mapWithArguments, namingConventions, nullSubstitution, preCondition, set, setMutate, setRecursiveValue, typeConverter, undefinedSubstitution };
^^^^^^

SyntaxError: Unexpected token 'export'
```

After manually adding `"type": "module"` to the package.json files of the installed `@automapper/core` and `@automapper/classes` packages everything works as expected.

I assume that without the module declaration the exports entries are ignored and the packages are imported as commonjs modules by default. I assume this affects all packages, but I only tested it with core and classes.

### Expected behavior

To import the packages as es modules.

### Package

- [ ] I don't know.
- [X] <code>@automapper/core</code>
- [X] <code>@automapper/classes</code>
- [ ] <code>@automapper/nestjs</code>
- [ ] <code>@automapper/pojos</code>
- [ ] <code>@automapper/mikro</code>
- [ ] <code>@automapper/sequelize</code>
- [ ] Other (see below)

### AutoMapper version

8.0.0

**Comment by nartc** [comment]
Hey thanks for reporting. What environment are you setting this up? I've tried several and thought it was working correctly.

**Comment by LennartH** [comment]
I'm using node v16.14.2 with typescript 4.6.3 as es module. 

This is the partial package.json:
```json
{
  "type": "module",
  "main": "./dist/index.js",
  "exports": {
    ".": "./dist/index.js",
    "./dto": "./dist/dto/index.js",
    "./model": "./dist/model/index.js",
    "./util": "./dist/util/index.js",
    "./schema.json": {
      "default": "./schema.json",
      "node": "./schema.json"
    },
    "./package.json": "./package.json"
  },
  "files": [
    "dist",
    "schema.json"
  ],
  "engines": {
    "node": "^12.20.0 || ^14.13.1 || >=16.0.0"
  },
  "scripts": {
    "build": "rimraf dist && npm run generate:json-schema && tsc -p tsconfig.build.json",
    "generate:json-schema": "node --no-warnings --unhandled-rejections=strict --loader ts-node/esm scripts/generate-json-schema.ts",
    "lint": "eslint --cache --cache-location .cache/.eslintcache --fix --ext .ts .",
    "test": "cross-env NODE_OPTIONS=--experimental-vm-modules NODE_NO_WARNINGS=1 jest",
    "prepare": "husky install"
  },
  "jest": {
    "preset": "ts-jest/presets/default-esm",
    "globals": {
      "ts-jest": {
        "useESM": true
      }
    },
    "transform": {
      "^.+\\.[tj]s$": "ts-jest"
    },
    "moduleNameMapper": {
      "^(\\.{1,2}/.*)\\.js$": "$1"
    },
    "setupFilesAfterEnv": [
      "jest-extended/all"
    ]
  }
}
```

This is the tsconfig.json:
```json
{
  "compilerOptions": {
    /* Basic Options */
    "target": "ES2020",
    "module": "ES2020",
    "outDir": "./dist",
    "sourceMap": true,
    "declaration": true,

    "strict": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "alwaysStrict": true,
    "strictPropertyInitialization": false,

    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,

    "moduleResolution": "Node",
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "forceConsistentCasingInFileNames": true,

    "emitDecoratorMetadata": true,
    "experimentalDecorators": true
  },
  "include": [
    "src",
    "test",
    "scripts"
  ],
  "exclude": [
    "dist",
    "node_modules"
  ]
}
```

Building the project with tsc worked fine but executing the `generate-json-schema` script resulted in the errors.

**Comment by nartc** [comment]
@LennartH Thanks. Is it possible that you could share a minimal reproduce?

**Comment by LennartH** [comment]
I'll try to set one up this evening.

**Comment by nartc** [comment]
@LennartH I was able to reproduce it with a Node ESM + TypeScript setup. This is painful. And honestly, I'm not entirely sure what's a correct approach moving forward. The ecosystem isn't ready for all-in in ESM 🥲. Any suggestion?

**Comment by LennartH** [comment]
Ok, good to know that it's not something weird in my specific setup :sweat_smile: ESM can be a real pain...

How are you currently testing if your modules work in different environments? My first attempt would be to add `"type": "module"` to all package.json files in packages. In the best case this would suffice since you declared how to import the packages in a commonjs context with the `exports` field.

**Comment by LennartH** [comment]
What about a setup described in the building section of https://www.sensedeep.com/blog/posts/2021/how-to-create-single-source-npm-module.html? The fixup script injecting the type field after the packages have been built feels a bit dirty, but I think it's easier to split the esm/commonjs building using two tsconfig files than creating separate `package.json` for each target package combination.

**Comment by nartc** [comment]
@LennartH Just published 8.1.0 which should fix this issue. I've tested with:

- NestJS (nodejs commonjs)
- Angular (ESM)
- Node TypeScript ESM boilerplate (node esm)

All I did was:
- Add `type:module` to all `package.json`
- Rename `index.cjs.js` to `index.cjs`

**Comment by LennartH** [comment]
Works good :+1: I'm glad the solution didn't require to change the building/publishing completely.
